alternator/executor.cc

```diff
@@ -1637,19 +1650,48 @@ future<executor::request_return_type> rmw_operation::execute(service::storage_pr
     });
 }
 
+bytes query_cache_key(const std::string_view& query, char query_type) {
+   auto b = md5_hasher::calculate(query);
+   b[0] = query_type; // this avoids key collision for different expression types
+   return b;
+}
+
+template<typename T>
+T parse_or_get_from_cache(lw_shared_ptr<expression_cache_t> cache, T parse_func(std::string_view), std::string_view query, char query_type) {
+    auto key = query_cache_key(query, query_type);
+    auto ptr = cache->find(key);
+    if (ptr) {
+        // we need to copy, it can be evicted from cache anytime and
+        // aditonally our expression evaluation code modifies T
+        return std::get<T>(*ptr);
+    }
+    // not found in cache, we need to put it there
+    auto loader_func = [parse_func, query_copy = std::string(query)] (bytes key) {
+        return make_ready_future<expression_cache_value_t>(parse_func(query_copy));
+    };
+    // leak future, we have no means to wait for it here
+    (void)cache->get_ptr(key, loader_func).discard_result()
+        .handle_exception([](auto ep) {/* ignore */}).then_wrapped([cache] (auto f) {
+        (void)cache; // keep it alive until we're done with it
+        return make_ready_future();
+    });
+    // parse it second time as synchronization with background task is not possible here
+    return parse_func(query);
```

Comment:
Perhaps we shouldn't be using loading_cache. Isn't everything synchronous?

Alternatively, return a future here, if it doesn't ruin the caller.

Reply:
Returning future here kind of ruins the caller, as non-futurized callstack is quite long, and it even goes through a constructor in some cases.

If there is some synchronous implementation I can also plug it in, I wanted to avoid writing my own cache, and also keep some similarity to cql code so it's potentially easier to maintain.

I imagine in the future we'd like to futurize more of alternator code, for similar reasons as having easily yieldable json handling code, then it would be easier to fix this place.

Reply:
Well, abandoning tasks like that isn't nice.

I understand not wanting to write a new cache, but reducing code quality also isn't good.

Reply:
We could enhance loading_cache with synchronous operations, but it's likely to be more difficult than writing a new cache.

Reply:
It's generally better to have one cache to have better utilization and bounded storage. 100 expressions per table -> 100,000 expressions if you have 1000 tables.